### PR TITLE
Added option for managed disks for data disks

### LINKED ISSUE #964: Add managed disk options to data disks
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]

Support optionally specifying ManagedDisk storage type for data disks https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/api/v1alpha3/types.go#L330:6, reusing the same type as OSDisks https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/api/v1alpha3/types.go#L323.

This is a good end-to-end feature for anyone looking to get familiar with CAPZ, and I'm happy to provide further guidance for anyone who wants to pick it up.

The solution should include:

- API changes 
- Update to the VM and VMSS services to set the options when creating the VM/VMSS
- Webhook validation
- Unit tests
- Template changes and/or documentation updates (data-disks.md)

Don't know where to start? Search "StorageAccountType" in the codebase to see how it's currently implemented for OS Disks.

/help
/good-first-issue

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- cluster-api-provider-azure version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by gdhuper**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Add support for ManagedDisk storage type for data disks.

**Which issue(s) this PR fixes**:
Fixes #964 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [x] API changes
- [x] Update to the VM and VMSS services to set the options when creating the VM/VMSS
- [x] Webhook Validation
- [x] Template changes and/or documentation updates (data-disks.md)
- [x] includes documentation
- [x] adds unit tests / e2e tests 
- [ ] squashed commits

**Release note**:

```release-note
Added option for managed disks for data disks.
```

### CODE REVIEW

**Comment by gdhuper** [comment]
I signed it

**Comment by gdhuper** [comment]
@CecileRobertMichon could you take a quick look and provide some feedback? Are there other changes that I might be missing besides unit tests?

File: api/v1alpha3/types.go [hunk 1]
```diff
@@ -376,6 +376,8 @@ type DataDisk struct {
 	NameSuffix string `json:"nameSuffix"`
 	// DiskSizeGB is the size in GB to assign to the data disk.
 	DiskSizeGB int32 `json:"diskSizeGB"`
+	// Managed disk option for a VM
+	ManagedDisk ManagedDisk `json:"managedDisk"`
```
Comment (CecileRobertMichon) [inline api/v1alpha3/types.go:380]: let's make this field optional (`// +optional`)  and `omitempty` so it is not breaking. By default, we want to keep the existing behavior.

File: api/v1alpha3/types.go [hunk 2]
```diff
@@ -376,6 +376,8 @@ type DataDisk struct {
 	NameSuffix string `json:"nameSuffix"`
 	// DiskSizeGB is the size in GB to assign to the data disk.
 	DiskSizeGB int32 `json:"diskSizeGB"`
+	// Managed disk option for a VM
```
Comment (CecileRobertMichon) [inline api/v1alpha3/types.go:379]: you'll need to run `make generate` to generate the updated CRDs

File: api/v1alpha3/azuremachine_validation.go [hunk 3]
```diff
@@ -94,6 +94,11 @@ func ValidateDataDisks(dataDisks []DataDisk, fieldPath *field.Path) field.ErrorL
 			nameSet[disk.NameSuffix] = struct{}{}
 		}
 
+		allErrs := append(allErrs, validateStorageAccountType(disk.ManagedDisk.StorageAccountType, fieldPath)...)
+
+		if errs:= ValidateManagedDisk(disk.ManagedDisk, disk.ManagedDisk, fieldPath.Child("managedDisk")); len(errs) > 0 {
```
Comment (CecileRobertMichon) [inline api/v1alpha3/azuremachine_validation.go:99]: let's unit test this

File: api/v1alpha3/azuremachine_validation.go [hunk 4]
```diff
@@ -165,6 +174,21 @@ func ValidateManagedDisk(old, new ManagedDisk, fieldPath *field.Path) field.Erro
 	return allErrs
 }
 
+// ValidateManagedDisks validates updates to Data disks' managed disk field.
+func ValidateManagedDisks(old, new []DataDisk, fieldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	for i, disk := range new {
```
Comment (CecileRobertMichon) [inline api/v1alpha3/azuremachine_validation.go:181]: here we should also check for len() of new and old so we don't run into an index out of range panic if they don't have equal lengths

**Comment by CecileRobertMichon** [comment]
@gdhuper please squash the commits

**Comment by CecileRobertMichon** [comment]
@nader-ziada this one modifies v1alpha3 types FYI

**Comment by nader-ziada** [comment]
> @nader-ziada this one modifies v1alpha3 types FYI

yeah, I saw that, will rebase on it when its merged

File: api/v1alpha3/azuremachine_validation.go (hunk 4 above)
Comment (gdhuper) [reply api/v1alpha3/azuremachine_validation.go:181]: If the len() mismatch, what is the error message that we want to return?

**Comment by gdhuper** [comment]
@CecileRobertMichon I ran into issues when squashing commits. Going to be closing out this PR and raise a new PR.

**Pull request closed without merging**